### PR TITLE
mergeKSortedLists.js

### PR DESCRIPTION
**Pull request by ganeshyevle**
You are given an array of 'K' linked-lists 'lists', each linked-list is sorted in ascending order. Merge all the linked lists into one sorted linkedlist and return it.

 Input 'lists =[[1,4,5],[1,3,4],[2,6]] 

Output ='[1,1,2,3,4,4,5,6]

### CODE REVIEW

File: DSA_Training/mergeKSortedLists.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+class ListNode {
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js:1]: please add comment

Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js:2]: Please add comment

File: DSA_Training/mergeKSortedLists.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    constructor(val = 0, next = null) {
+        this.val = val;      // Value of the node
+        this.next = next;    // Pointer to the next node
+    }
+}
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js:8]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        tail = tail.next;            // Move tail pointer to new end
+        
+        if (minNode.next) {
+            minHeap.push(minNode.next); // Push next node of minNode to heap
+        }
+    }
+    
+    return dummy.next; // Return head of merged sorted list
+}
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js:36]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+class MinHeap {
+    constructor() {
+        this.heap = []; // Array to store nodes forming the heap
+    }
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        
+        if (minNode.next) {
+            minHeap.push(minNode.next); // Push next node of minNode to heap
+        }
+    }
+    
+    return dummy.next; // Return head of merged sorted list
+}
+
+class MinHeap {
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js:38]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    constructor() {
+        this.heap = []; // Array to store nodes forming the heap
+    }
+    
+    push(node) {
+        this.heap.push(node); // Add node to heap
+        this.heapifyUp(this.heap.length - 1); // Maintain heap property
+    }
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        const lastNode = this.heap.pop(); // Remove last node from heap
+        
+        if (this.heap.length > 0) {
+            this.heap[0] = lastNode; // Move last node to root
+            this.heapifyDown(0);     // Maintain heap property
+        }
+        
+        return minNode; // Return smallest node
+    }
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            this.heapifyDown(0);     // Maintain heap property
+        }
+        
+        return minNode; // Return smallest node
+    }
+    
+    isEmpty() {
+        return this.heap.length === 0; // Check if heap is empty
+    }
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    heapifyUp(index) {
+        let parent = Math.floor((index - 1) / 2); // Parent index
+        
+        while (index > 0 && this.heap[index].val < this.heap[parent].val) {
+            this.swap(index, parent); // Swap if current node is smaller than parent
+            index = parent;           // Move index to parent
+            parent = Math.floor((index - 1) / 2); // Update parent index
+        }
+    }
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js]: Please add comment here

File: DSA_Training/mergeKSortedLists.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        if (right < this.heap.length && this.heap[right].val < this.heap[smallest].val) {
+            smallest = right; // Update smallest if right child is smaller
+        }
+        
+        if (smallest !== index) {
+            this.swap(index, smallest); // Swap and heapify down recursively
+            this.heapifyDown(smallest);
+        }
+    }
+
```
Comment (Umarfarook1) [inline DSA_Training/mergeKSortedLists.js]: Please add comment here

**Review by Umarfarook1 — CHANGES_REQUESTED** [review]
Please go through requested changes

File: DSA_Training/mergeKSortedLists.js (hunk 1 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js:2]: done

File: DSA_Training/mergeKSortedLists.js (hunk 2 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js:8]: done

File: DSA_Training/mergeKSortedLists.js (hunk 3 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js:36]: done

File: DSA_Training/mergeKSortedLists.js (hunk 4 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js]: done

File: DSA_Training/mergeKSortedLists.js (hunk 5 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js:38]: done

File: DSA_Training/mergeKSortedLists.js (hunk 6 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js]: done

File: DSA_Training/mergeKSortedLists.js (hunk 7 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js]: done

File: DSA_Training/mergeKSortedLists.js (hunk 8 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js]: done

File: DSA_Training/mergeKSortedLists.js (hunk 9 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js]: done

File: DSA_Training/mergeKSortedLists.js (hunk 10 above)
Comment (ganeshyevle) [reply DSA_Training/mergeKSortedLists.js]: done

**Pull request merged**